src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc

```diff
@@ -276,15 +317,51 @@ std::string ChooseServiceConfig(char* service_config_choice_json,
   return service_config->Dump();
 }
 
-void AresClientChannelDNSResolver::AresRequestWrapper::OnResolved(
+void AresClientChannelDNSResolver::AresRequestWrapper::OnHostnameResolved(
+    void* arg, grpc_error_handle error) {
+  auto* self = static_cast<AresRequestWrapper*>(arg);
+  {
+    MutexLock lock(&self->on_resolved_mu_);
+    self->hostname_request_.reset();
+    self->OnResolvedLocked(error);
+  }
+  self->Unref(DEBUG_LOCATION, "OnHostnameResolved");
+}
+
+void AresClientChannelDNSResolver::AresRequestWrapper::OnSRVResolved(
+    void* arg, grpc_error_handle error) {
+  auto* self = static_cast<AresRequestWrapper*>(arg);
+  {
+    MutexLock lock(&self->on_resolved_mu_);
+    self->srv_request_.reset();
+    self->OnResolvedLocked(error);
+  }
+  self->Unref(DEBUG_LOCATION, "OnSRVResolved");
+}
+
+void AresClientChannelDNSResolver::AresRequestWrapper::OnTXTResolved(
     void* arg, grpc_error_handle error) {
   auto* self = static_cast<AresRequestWrapper*>(arg);
-  self->OnResolved(error);
+  {
+    MutexLock lock(&self->on_resolved_mu_);
+    self->txt_request_.reset();
+    self->OnResolvedLocked(error);
+  }
+  self->Unref(DEBUG_LOCATION, "OnTXTResolved");
 }
 
-void AresClientChannelDNSResolver::AresRequestWrapper::OnResolved(
-    grpc_error_handle error) {
-  GRPC_CARES_TRACE_LOG("resolver:%p OnResolved()", this);
+void AresClientChannelDNSResolver::AresRequestWrapper::OnResolvedLocked(
+    grpc_error_handle error) ABSL_EXCLUSIVE_LOCKS_REQUIRED(on_resolved_mu_) {
+  if (hostname_request_ != nullptr || srv_request_ != nullptr ||
+      txt_request_ != nullptr) {
+    GRPC_CARES_TRACE_LOG(
```

Comment:
Should this log message be outside of this block?  As currently written, it will be logged only if all three requests are null, so the "waiting" messages will never be logged.